### PR TITLE
Fix xamarin deadlocks

### PR DESCRIPTION
**Pull request by RomaRudyak**
This pull request contains changes to Connect ans SessionLoop method on XmppConnection.

In XmppTcpConnection used tasks for making connections requests and returning from ConnectAsync when session created or Connection fails.

Add test Xamarin.Forms project with basic implementation connection to xmpp server.

You can have concern about this one:
Made SASLSCRAM as default SASL handler in order that free servers not return any Available mechanisms

### CODE REVIEW

**Comment by vitalyster** [comment]
What is the server software on chat.chinwag.im and why it violate specs? It must send stream features

**Comment by vitalyster** [comment]
I'm connect to `chinwag.im` (note, not `chat.chinwag.im`!) and got correct response:
```
<stream:features xmlns:stream="http://etherx.jabber.org/streams">
  <register xmlns="http://jabber.org/features/iq-register" />
  <mechanisms xmlns="urn:ietf:params:xml:ns:xmpp-sasl">
    <mechanism>PLAIN</mechanism>
    <mechanism>SCRAM-SHA-1</mechanism>
    <mechanism>SCRAM-SHA-1-PLUS</mechanism>
  </mechanisms>
</stream:features>
```
So SASL changes are not needed

**Comment by vitalyster** [comment]
I have fixed starttls connection in https://github.com/vitalyster/SharpXMPP/commit/17be7ab582f5f5b46d7369630cb4572ab9a2684f
Please rebase your Xamarin client and drop SASL changes, it should work now. Let me know if anything wrong

**Comment by vitalyster** [comment]
Also, please move Xamarin projects to SharpXMPP.Xamarin in order to match existing structure

File: SharpXMPP.Shared/XmppTcpConnection.cs [hunk 1]
```diff
@@ -121,34 +144,128 @@ public override void SessionLoop()
             }
         }
 
-        public override async void Connect()
+        // For backward compatibility
+        public override async void Connect() => await ConnectAsync();
```
Comment (ForNeVeR) [inline SharpXMPP.Shared/XmppTcpConnection.cs]: `async void` is a terrible solution and won't keep any "compatibility" for us. Let's remove that, please.

I don't think that keeping backward compatibility is that big issue for `v0.0.1` currently, but we need to discuss that with @vitalyster.

File: SharpXMPP.sln [hunk 2]
```diff
@@ -19,8 +19,26 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "SharpXMPP.NUnit", "SharpXMP
...
+		Ad-Hoc|Any CPU = Ad-Hoc|Any CPU
+		Ad-Hoc|iPhone = Ad-Hoc|iPhone
+		Ad-Hoc|iPhoneSimulator = Ad-Hoc|iPhoneSimulator
+		Ad-Hoc|Mixed Platforms = Ad-Hoc|Mixed Platforms
+		Ad-Hoc|x86 = Ad-Hoc|x86
+		AppStore|Any CPU = AppStore|Any CPU
+		AppStore|iPhone = AppStore|iPhone
+		AppStore|iPhoneSimulator = AppStore|iPhoneSimulator
+		AppStore|Mixed Platforms = AppStore|Mixed Platforms
+		AppStore|x86 = AppStore|x86
```
Comment (ForNeVeR) [inline SharpXMPP.sln]: Do we really need all these strange configurations? At lease get rid of `Mixed Platforms`, it seems that's redundant.

**Comment by ForNeVeR** [comment]
Just to be clear: you've implemented both iOS and Android client apps here, right?

File: SharpXMPP.Shared/XmppTcpConnection.cs (hunk 1 above)
Comment (vitalyster) [reply SharpXMPP.Shared/XmppTcpConnection.cs]: I'm agree to remove

File: SharpXMPP.Shared/XmppTcpConnection.cs [hunk 3]
```diff
@@ -88,11 +91,31 @@ public override void Send(XElement data)
...
             Writer.WriteEndElement();
         }
 
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing && !_disposed)
+            {
+                _disposed = true;
+                // NOTE: used this statement because faced issue with compilation under net451 
+                (_client as IDisposable)?.Dispose();
```
[the remark this answers is not in the document]
Comment (RomaRudyak) [reply SharpXMPP.Shared/XmppTcpConnection.cs]: Close() implemented only for net451. In netstandart1.6 no Close method.

We can think about moving to netstandart2.0 but as this is a library it batter to target the lowest possible.

File: SharpXMPP.sln (hunk 2 above)
Comment (RomaRudyak) [reply SharpXMPP.sln]: Ok

**Comment by RomaRudyak** [comment]
Yes. But for iOS I need add initialization of Xamarin.Forms.

File: SharpXMPP.Shared/XmppConnection.cs [hunk 4]
```diff
@@ -26,10 +28,25 @@ public class SignedInArgs
...
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        ~XmppConnection()
+        {
+            Dispose(false);
+        }
```
[the remark this answers is not in the document]
Comment (RomaRudyak) [reply SharpXMPP.Shared/XmppConnection.cs]: Ok, I will remove finalizer and leave  Dispose

File: SharpXMPP.Shared/XmppTcpConnection.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ForNeVeR) [reply SharpXMPP.Shared/XmppTcpConnection.cs]: Oh heck, okay then. Let's then replace that with a direct cast `((IDisposable)_client)?.Dispose()`. We know for fact that the cast should succeed for any non-null `_client`.

**Comment by ForNeVeR** [comment]
And, yeah, thanks a lot for your contribution!

**Comment by vitalyster** [comment]
Please rebase this branch onto master and force push in order to resolve conflicts

**Comment by vitalyster** [comment]
My fix was: re-read stream features after TLS handshake, I didn't see this in current code

**Comment by RomaRudyak** [comment]
This code in ac1f51b commit.
After tls enabled method line 159 in XmppTcpConnection.cs

**Comment by ForNeVeR** [comment]
@vitalyster it's okay, the pull request is now rebased.

**Comment by vitalyster** [comment]
Ok, just tested and it works for me. @ForNeVeR, should we wait for your review?

**Comment by ForNeVeR** [comment]
@vitalyster yep, wait please. ETA 2.5 hours.

**Comment by vitalyster** [comment]
We also need to fix appveyour build: either disable iOS/Android projects or configure SDKs properly

**Comment by RomaRudyak** [comment]
issue relates to package restoring.

in appveyor used `dotnet restore` which cant restore packages for Android/iOS project but when call `nuget restore` all will be good.

**Comment by ForNeVeR** [comment]
Alright, could you add `nuget restore` to the `before_build` script? [Here's an example](https://github.com/ForNeVeR/SvgBuild/blob/ecdb0b1d13856a91ec92aa6b9d7f784c51e6881f/appveyor.yml#L5-L6).

**Comment by ForNeVeR** [comment]
Need to fix Appveyor build now, and ok to merge then.

**Comment by ForNeVeR** [comment]
Excellent, thanks a lot!

@vitalyster, feel free to merge.

**Pull request merged**